provider/bigquery.go

```diff
@@ -0,0 +1,1377 @@
+	setTablePrefix(prefix string)
+	getContext() context.Context
+	setContext()
+	castTableItemType(v interface{}, t interface{}) interface{}
+	materializationExists(tableName string) string
+	materializationDrop(tableName string) string
+	materializationUpdate(client *bigquery.Client, tableName string, sourceName string) error
+	monitorJob(job *bigquery.Job) error
+	transformationCreate(name string, query string) string
+	getColumns(client *bigquery.Client, name string) ([]TableColumn, error)
+	transformationUpdate(client *bigquery.Client, tableName string, query string) error
+	trainingSetCreate(store *bqOfflineStore, def TrainingSetDef, tableName string, labelName string) error
+	trainingSetUpdate(store *bqOfflineStore, def TrainingSetDef, tableName string, labelName string) error
+	trainingSetQuery(store *bqOfflineStore, def TrainingSetDef, tableName string, labelName string, isUpdate bool) error
+	atomicUpdate(client *bigquery.Client, tableName string, tempName string, query string) error
+	trainingRowSelect(columns string, trainingSetName string) string
+	primaryTableRegister(tableName string, sourceName string) string
+}
+
+type defaultBQQueries struct {
+	TablePrefix string
+	Ctx         context.Context
+}
+
+type bqGenericTableIterator struct {
+	iter         *bigquery.RowIterator
+	currentValue GenericRecord
+	err          error
+	query        BQOfflineTableQueries
+}
+
+type bqPrimaryTable struct {
+	client *bigquery.Client
+	name   string
+	query  BQOfflineTableQueries
+	schema TableSchema
+}
+
+func (pt *bqPrimaryTable) GetName() string {
+	return pt.name
+}
+
+func (pt *bqPrimaryTable) IterateSegment(n int64) (GenericTableIterator, error) {
+	tableName := fmt.Sprintf("%s.%s", pt.query.getTablePrefix(), pt.name)
+	query := fmt.Sprintf("SELECT * FROM `%s` LIMIT %d", tableName, n)
+	bqQ := pt.client.Query(query)
+	it, err := bqQ.Read(pt.query.getContext())
+
+	if err != nil {
+		return nil, err
+	}
+	return newBigQueryGenericTableIterator(it, pt.query), nil
+}
+
+func (pt *bqPrimaryTable) NumRows() (int64, error) {
+	var n []bigquery.Value
+	tableName := fmt.Sprintf("%s.%s", pt.query.getTablePrefix(), pt.name)
+	query := fmt.Sprintf("SELECT COUNT(*) FROM `%s`", tableName)
+
+	bqQ := pt.client.Query(query)
+
+	it, err := bqQ.Read(pt.query.getContext())
+	if err != nil {
+		return 0, err
+	}
+
+	err = it.Next(&n)
+	if err != nil {
+		return 0, err
+	}
+
+	return n[0].(int64), nil
+}
+
+func (pt *bqPrimaryTable) Write(rec GenericRecord) error {
+	tb := pt.name
+	recordsParameter, columns, columnsString := pt.getNonNullRecords(rec)
+
+	placeholder := pt.query.createValuePlaceholderString(columns)
+	upsertQuery := pt.query.upsertQuery(tb, columnsString, placeholder)
+
+	bqQ := pt.client.Query(upsertQuery)
+	bqQ.Parameters = recordsParameter
+
+	_, err := bqQ.Read(pt.query.getContext())
+
+	return err
+}
+
+func (pt *bqPrimaryTable) getNonNullRecords(rec GenericRecord) ([]bigquery.QueryParameter, []TableColumn, string) {
```

Comment:
Whats this function do?

Reply:
Basically, I couldn't add `<nil>` values so I made it that I add the non-null values and the columns with nulls will be automatically set to null.

Reply:
Is it the driver not accepting nil values as an acceptable value or something?

Reply:
the driver doesn't accept null values in the current query approach. i.e. ```INSERT INTO table (col1) VALUES (nil)```